include/nbl/builtin/hlsl/subgroup/arithmetic_portability_impl.hlsl

```diff
@@ -0,0 +1,490 @@
+// so we implement them distinctly
+
+[[vk::ext_instruction(353)]]
+int spirv_subgroupPrefixMin(uint scope, [[vk::ext_literal]] uint operation, int value);
+[[vk::ext_instruction(354)]]
+uint spirv_subgroupPrefixMin(uint scope, [[vk::ext_literal]] uint operation, uint value);
+[[vk::ext_instruction(355)]]
+float spirv_subgroupPrefixMin(uint scope, [[vk::ext_literal]] uint operation, float value);
+
+template<>
+struct inclusive_scan<int, binops::min<int> >
+{
+    int operator()(const int x)
+    {
+        return spirv_subgroupPrefixMin(3, 1, x);
+    }
+};
+
+template<>
+struct inclusive_scan<uint, binops::min<uint> >
+{
+    uint operator()(const uint x)
+    {
+        return spirv_subgroupPrefixMin(3, 1, x);
+    }
+};
+
+template<>
+struct inclusive_scan<uint, binops::min<float> >
+{
+    float operator()(const float x)
+    {
+        return spirv_subgroupPrefixMin(3, 1, x);
+    }
+};
+
+template<>
+struct exclusive_scan<int, binops::min<int> >
+{
+    int operator()(const int x)
+    {
+        return spirv_subgroupPrefixMin(3, 2, x);
+    }
+};
+
+template<>
+struct exclusive_scan<uint, binops::min<uint> >
+{
+    uint operator()(const uint x)
+    {
+        return spirv_subgroupPrefixMin(3, 2, x);
+    }
+};
+
+template<>
+struct exclusive_scan<uint, binops::min<float> >
+{
+    float operator()(const float x)
+    {
+        return spirv_subgroupPrefixMin(3, 2, x);
+    }
+};
+
+// *** MAX ***
+template<typename T>
+struct reduction<T, binops::max<T> >
+{
+    T operator()(const T x)
+    {
+        return WaveActiveMax(x);
+    }
+};
+
+[[vk::ext_instruction(356)]]
+int spirv_subgroupPrefixMax(uint scope, [[vk::ext_literal]] uint operation, int value);
+[[vk::ext_instruction(357)]]
+uint spirv_subgroupPrefixMax(uint scope, [[vk::ext_literal]] uint operation, uint value);
+[[vk::ext_instruction(358)]]
+float spirv_subgroupPrefixMax(uint scope, [[vk::ext_literal]] uint operation, float value);
+
+template<>
+struct inclusive_scan<int, binops::max<int> >
+{
+    int operator()(const int x)
+    {
+        return spirv_subgroupPrefixMax(3, 1, x);
+    }
+};
+
+template<>
+struct inclusive_scan<uint, binops::max<uint> >
+{
+    uint operator()(const uint x)
+    {
+        return spirv_subgroupPrefixMax(3, 1, x);
+    }
+};
+
+template<>
+struct inclusive_scan<uint, binops::max<float> >
+{
+    float operator()(const float x)
+    {
+        return spirv_subgroupPrefixMax(3, 1, x);
+    }
+};
+
+template<>
+struct exclusive_scan<int, binops::max<int> >
+{
+    int operator()(const int x)
+    {
+        return spirv_subgroupPrefixMax(3, 2, x);
+    }
+};
+
+template<>
+struct exclusive_scan<uint, binops::max<uint> >
+{
+    uint operator()(const uint x)
+    {
+        return spirv_subgroupPrefixMax(3, 2, x);
+    }
+};
+
+template<>
+struct exclusive_scan<uint, binops::max<float> >
+{
+    float operator()(const float x)
+    {
+        return spirv_subgroupPrefixMax(3, 2, x);
+    }
+};
+
+}
+#else
+namespace portability
+{
```

Comment:
but the `portability` version and anything that might use it, needs a HUGE warning present in the GLSL version

Namely shuffles where you attempt to read an invactive lane, return garbage, which means that our portability reductions and prefix sums will also return garbage/UB/UV

Whereas the native instructions would substitute any inactive lane for the identity value and still work.